### ISSUE TITLE
Trait default implementations

### ISSUE THREAD

**Issue by cars10** [antoyo/relm#66]
Is there any reason why there are no default implementations for the trait methods available?
It's kinda annoying to write for example an empty `update` method when my widget does not even have a msg enum or an associated model.

Empty default implementations would greatly reduce boilerplate, at least in my case.

**Comment by antoyo** [comment]
The `update` method does not have a default implementation because it used to return a value.
Now that it does not return one anymore, I could make an empty default implementation.
Do you need a default implementation for other methods?
If so, what would be the return value of these default implementations?

**Comment by cars10** [comment]
Specifically:

* Trait `Update`. Methods: `model`, `#update`
* Trait `Container` Method: `other_containers`

These are the ones that are empty in most cases on my codebase - specifically in this combination. I have a rather complex UI with many containers and deep nesting, and most container widgets do not have a model or msg.

**Comment by antoyo** [comment]
The thing with the `model` method is that it should return a value.
Unless I had a trait bound like `Self::Model: Default` which I don't want to, I cannot create a default implementation.
What I could do though, is to automatically create this method if it's not provided when using the `#[widget]` attribute, which is what I do for other methods.

For the `other_containers` method, there's the same issue: we need to return a value.
However, in this case, if you use the `#[widget]` attribute, this method will be automatically generated.

So, would it be okay if I make a default implementation for `Update::update()` and auto-generate the `Update::model` method when using the `#[widget]` if missing?

**Comment by cars10** [comment]
Wouldn't it be ok to do something like this?

```rust
fn model(_: &Relm<Self>, _: ()) -> () {}
```

```rust
fn other_containers(&self) -> () {}
```

Thats what i write right now - as i do not have a model, i return `()` .

**Comment by antoyo** [comment]
No, it works on your case, but since `Update::model` returns `Self::Model` in this type can be anything, I cannot write the default implementation in a generic way without adding a trait bound like `Default`.
Same for the other method.

Why don't you use the `#[widget]` attribute?
You know there's a way to use it on stable Rust?

**Comment by cars10** [comment]
Ah, makes sense. 

I didn't try the `#[widget]` attribute yet. The long list of warnings in the readme had me think that i will be better off with not using it.

**Comment by antoyo** [comment]
So do you need the proposed changes?
I'll probably make them later if you don't need them.
